### PR TITLE
Disable autocast cache when grad is disabled

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113685

Fixes https://github.com/pytorch/pytorch/issues/65766 and https://github.com/pytorch/pytorch/issues/112583

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by albanD — APPROVED** [review]
Thanks

**Comment by ezyang** [comment]
Hmm, this test failure is kind of interesting. It is failing because the cache access is happening inside of an autograd.Function, at which point in time autograd has been disabled (because we're going to set grad_fn after we get out of the cast.) I guess we'll have to do something more complicated.

**Comment by ezyang** [comment]
A more complicated version, with comments. Please rereview @albanD @Skylion007

File: aten/src/ATen/autocast_mode.cpp [hunk 1]
```diff
@@ -210,7 +210,30 @@ Tensor cached_cast(at::ScalarType to_type, const Tensor& arg, DeviceType device_
...
+        //
+        // Now, you might ask, why don't we test if
+        // c10::GradMode::is_enabled() before asking if we can_try_cache?
+        // Sometimes we are doing autocasting in a context where we will
+        // eventually attach a grad_fn to it later (e.g., custom autograd
+        // function).  In this case, the autocast will occur in a context
+        // where grad is disabled... but eventually everything will work out
+        // (the cached entry will requires_grad.)  This case is explicitly
+        // tested for, so I decided to go this more circuitous route.
+        if (!cand_casted_arg.requires_grad() && c10::GradMode::is_enabled()) {
```
Comment (albanD) [inline aten/src/ATen/autocast_mode.cpp:230]:
```suggestion
        if (!cand_casted_arg.requires_grad() && c10::GradMode::is_enabled() && arg.requires_grad()) {
```

File: aten/src/ATen/autocast_mode.cpp [hunk 2]
```diff
@@ -210,7 +210,30 @@ Tensor cached_cast(at::ScalarType to_type, const Tensor& arg, DeviceType device_
...
+          auto casted_arg = arg.to(to_type);
+          cached_casts.emplace(arg.unsafeGetTensorImpl(), val_type{weakref_type(arg.getIntrusivePtr()), casted_arg});
```
Comment (albanD) [inline aten/src/ATen/autocast_mode.cpp:232]: duplication is meh. But not easy to avoid I agree.

File: aten/src/ATen/autocast_mode.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/autocast_mode.cpp:230]: arg.requires_grad() is guaranteed by the `can_try_cache` test above.

**Comment by ezyang** [comment]
this PR is probably dead unless someone helps figure out the transformer encoder problem. cc @drisspg maybe

**Comment by drisspg** [comment]
This seems like this might be real? since the code seems to be exercising your new logic.
That being said the actual numerical change is pretyy small ..

``` Python
        with maybe_autocast:
            for input, src_key_padding_mask in input_mask_pairs:
                with torch.no_grad():
                    fastpath_output = model(input, src_key_padding_mask=src_key_padding_mask)
                    
```

**Pull request closed without merging**